text/0000-os-str-pattern.md

```diff
@@ -0,0 +1,527 @@
+`"\u{10000}"[3..]`), a panic will be raised, similar to that of `str`. The following are guaranteed
+to be valid positions on all platforms:
+
+* `0`.
+* `self.len()`.
+* The returned indices from `find()`, `rfind()`, `match_indices()` and `rmatch_indices()`.
+* The returned ranges from `find_range()`, `rfind_range()`, `match_ranges()` and `rmatch_ranges()`.
+
+Index arithmetic is wrong for `OsStr`, i.e. `i + n` may not produce the correct index (see
+[Drawbacks](#drawbacks)).
+
+For WTF-8 encoding on Windows, we define:
+
+* boundary of a character or surrogate byte sequence is Valid.
+* middle (byte 2) of a 4-byte sequence is Valid.
+* interior of a 2- or 3-byte sequence is Invalid.
+* byte 1 or 3 of a 4-byte sequence is Invalid.
+
+Outside of Windows where the `OsStr` consists of arbitrary bytes, all numbers within
+`0 ..= self.len()` are considered a valid index. This is because we want to allow
+`os_str.find(OsStr::from_bytes(b"\xff"))`, and thus cannot use UTF-8 to reason with a Unix `OsStr`.
+
+Note that we have never guaranteed the actual `OsStr` encoding, these should only be considered an
+implementation detail.
+
+## Comparison and storage
+
+All `OsStr` strings with sliced 4-byte sequence can be converted back to proper WTF-8 with an O(1)
+transformation:
+
+* If the string starts with `[\x80-\xbf]{3}`, replace these 3 bytes with the canonical low surrogate
+    encoding.
+* If the string ends with `[\xf0-\xf4][\x80-\xbf]{2}`, replace these 3 bytes with the canonical high
+    surrogate encoding.
+
+We can this transformation “*canonicalization*”.
+
+All owned `OsStr` should be canonicalized to contain well-formed WTF-8 only: `Box<OsStr>`,
+`Rc<OsStr>`, `Arc<OsStr>` and `OsString`.
+
+Two `OsStr` are compared equal if they have the same canonicalization. This may slightly reduce the
+performance with a constant overhead, since there would be more checking involving the first and
+last three bytes.
+
+## Matching
+
+When an `OsStr` is used for matching, an unpaired low surrogate at the beginning and unpaired high
+surrogate at the end must be replaced by regular expressions that match all pre-canonicalization
+possibilities. For instance, matching for `xxxx\u{d9ab}` would create the following regex:
+
+```
+xxxx(
+    \xed\xa6\xab        # canonical representation
+|
+    \xf2\x86[\xb0-\xbf] # split representation
+)
+```
+
+and matching for `\u{dcef}xxxx` with create the following regex:
+
+```
+(
+    \xed\xb3\xaf                        # canonical representation
+|
+    [\x80-\xbf][\x83\x93\xa3\xb3]\xaf   # split representation
+)xxxx
+```
+
+After finding a match, if the end points to the middle of a 4-byte sequence, the search engine
+should move backward by 2 bytes before continuing. This ensure searching for `\u{dc00}\u{d800}` in
+`\u{10000}\u{10000}\u{10000}` will properly yield 2 matches.
+
+## Pattern API
+
+This RFC assumes a generalized pattern API which supports more than strings. If the pattern API is
```

Comment:
I don’t quite understand what it means for an RFC to assume something. I’d prefer this RFC to propose one concrete API for `Pattern` and related traits (and maybe list alternatives in the Alternatives section).

Reply:
I've expanded this section and linked to the previous drafts. However, I prefer not to formally propose the pattern API here in this RFC as that is kinda out-of-scope.